src/checkpoint.rs

```diff
@@ -0,0 +1,141 @@
+use crate::errors::Result;
+use crate::mock::MockStorage;
+use crate::traits::{Api, Extern, ReadonlyStorage, Storage};
+
+pub struct Checkpoint<'a, S: Storage> {
+    /// a backing storage that is only modified upon commit
+    storage: &'a mut S,
+    /// these are local changes not flushed to backing storage
+    local_state: MockStorage,
+    /// this is a list of changes to be written to backing storage upon commit
+    rep_log: Vec<Op>,
+}
+
+enum Op {
+    Set { key: Vec<u8>, value: Vec<u8> },
+}
+
+impl<'a, S: Storage> Checkpoint<'a, S> {
+    pub fn new(storage: &'a mut S) -> Self {
+        Checkpoint {
+            storage,
+            local_state: MockStorage::new(),
```

Comment:
shouldn't this make a copy of `storage`'s content?

Reply:
No, it is zero copy.
It overlays local changes on an immutible backing store (that is only written on commit).

No way I'm going to copy a production database here. This is the same logic used in cosmos-sdk and weave to handle potentially roll-backable modifications.

I believe it is alternately called `cache` and `savepoint` there

Reply:
I checked the implementation of `.get` again. Makes sense now.